lib/Dialect/Random/IR/RandomOps.td

```diff
@@ -0,0 +1,105 @@
+    protocols that agree on shared randomness. The PRNG is used to initialized
+    the random distributions such as the discrete gaussian distribution and the
+    discrete uniform distribution. This initialization also takes as input a
+    number of bits that are generated for each number value sampled (num_bits).
+    For instance, a num_bits of 32 will mean that distributions will generate a
+    32-bit integer value. We expect that the seed initialization is done statically
+    and globally once per thread for all distributions; however, if multiple threads are
+    generating randomness, then seed initialization should be done per thread;
+    otherwise there is no guarantee of consistent behavior. Thread safety is so
+    far not considered.
+  }];
+
+  let arguments = (ins
+    SignlessIntegerLike:$seed,
+    Builtin_IntegerAttr:$num_bits
+  );
+  let results = (outs PRNG:$output);
+}
+
+def Random_DiscreteGaussianDistributionOp : Random_Op<"discrete_gaussian_distribution", [Pure]> {
+  let summary = "Initializes the Discrete Gaussian Distribution";
+  let description = [{
+    Initializes the Discrete Gaussian Distribution. The distribution is
+    initialized with a mean and a standard deviation and pseudorandom generator
+    that provides the source of the randomness.
+  }];
+
+  let arguments = (ins
+    PRNG:$input,
+    Builtin_IntegerAttr:$mean,
+    ConfinedAttr<Builtin_IntegerAttr, [IntNonNegative]>:$stddev
+  );
+  let results = (outs Distribution:$output);
+}
+
+def Random_DiscreteUniformDistributionOp: Random_Op<"discrete_uniform_distribution", [Pure]> {
+  let summary = "Initializes the Discrete Uniform Distribution";
+  let description = [{
+    Initializes the Discrete Uniform Distribution. The distribution is
+    initialized with a minimum and a maximum value and pseudo random generator
+    that provides the source of the randomness. The distribution is inclusive of
+    the minimum and exclusive of the maximum.
+  }];
+
+  let arguments = (ins
+    PRNG:$input,
+    Builtin_IntegerAttr:$min,
+    Builtin_IntegerAttr:$max
+  );
+  let results = (outs Distribution:$output);
+  let assemblyFormat = "$input `{` `range` `=` `[` $min `,` $max `]` `}` attr-dict `:` `(` qualified(type($input)) `)` `->` type($output)";
+  let hasVerifier = 1;
+}
+
+def Random_SampleOp : Random_Op<"sample"> {
+  let summary = "Samples from a distribution";
+  let description = [{
+    Samples from the distribution to obtain a random value.
+  }];
+
+  let arguments = (ins
+    Distribution:$input
+  );
+  let results = (outs AnySignlessInteger:$output);
+}
+
+def Random_SampleTensorOp : Random_Op<"sample_tensor"> {
+  let summary = "Samples a tensor of random values from a distribution";
+  let description = [{
+    Samples a tensor of random values by sampling each element from the provided
+    distribution.
+  }];
+
+  let arguments = (ins
+    Distribution:$input,
+    DenseI32ArrayAttr:$shape
```

Comment:
This shape may not even be needed. The op already expresses this shape by the shape of the result, which can be thought of as a "parameter" to the op (same way arith.constant specifies the type of the result as an "input" to the op).

In which case, I'm not so sure this op needs to be separated from `sample`. It could just be a single op whose result may be either a signless integer or a tensor thereof. I.e., reuse `SignlessIntegerLike` from `mlir/IR/CommonTypeConstraints.td`. Passes would have to branch on the type of the result, but for this op it seems relatively harmless.

Reply:
```
def Random_SampleOp : Random_Op<"sample"> {
  let summary = "Samples from a distribution";
  let description = [{
    Samples from the distribution to obtain a random value
    or tensor of values.
  }];

  let arguments = (ins
    Distribution:$input
  );
  let results = (outs SignlessIntegerLike:$output);
}
```